Assets/Src/Shaders/GridShader.shader

```diff
@@ -0,0 +1,180 @@
+
+    // The alpha channel of _GridColor defines how much the grid color is blended into the main
+    // texture.
+    _GridColor("Grid Color", Color) = (1, 1, 1, 1)
+    // _GridSize indicates the cell number along each axis in the uv space. The primary grid will be
+    // turned off if it is set to zero.
+    _GridSize("Grid Size", Range(0.0, 30.0)) = 10.0
+    // The line width of the primary grid.
+    _GridLineWidth("Grid Line Width", Range(1.0, 50.0)) = 2.0
+
+    // The customizable properties of the secondary grid.
+
+    // The alpha channel of _SecondaryGridColor defines how much the secondary grid color is blended
+    // into the main texture.
+    _SecondaryGridColor("Secondary Grid Color", Color) = (1, 1, 1, 1)
+    // _SecondaryGridSize indicates the cell number along each axis inside a primary grid cell. The
+    // secondary grid will be turned off if it is set to zero.
+    _SecondaryGridSize("Secondary Grid Size", Range(0.0, 10.0)) = 0.0
+    // The line width of the secondary grid.
+    _SecondaryGridLineWidth("Secondary Grid Line Width", Range(1.0, 50.0)) = 1.0
+  }
+
+  SubShader {
+    Tags {
+      "Queue" = "Transparent"
+      "IgnoreProjector" = "True"
+      "RenderType" = "Transparent"
+    }
+    ZWrite Off
+    Blend SrcAlpha OneMinusSrcAlpha
+    LOD 100
+
+    Pass {
+      CGPROGRAM
+      #pragma vertex vert
+      #pragma fragment frag
+
+      #include "UnityCG.cginc"
+      #include "Lighting.cginc"
+
+      struct a2v {
+        float4 pos : POSITION;
+        float3 normal : NORMAL;
+        float2 uv : TEXCOORD0;
+      };
+
+      struct v2f {
+        float4 pos : SV_POSITION;
+        float3 worldNormal : TEXCOORD0;
+        float3 worldPos : TEXCOORD1;
+        float2 uv : TEXCOORD2;
+      };
+
+      sampler2D _MainTex;
+      float4 _MainTex_ST;
+      fixed4 _Color;
+      fixed4 _Specular;
+      float _Gloss;
+      fixed4 _GridColor;
+      float _GridLineWidth;
+      float _GridSize;
+      fixed4 _SecondaryGridColor;
+      float _SecondaryGridLineWidth;
+      float _SecondaryGridSize;
+
+      v2f vert(a2v v) {
+        v2f o;
+        o.pos = UnityObjectToClipPos(v.pos);
+        o.worldNormal = UnityObjectToWorldNormal(v.normal);
+        o.worldPos = mul(unity_ObjectToWorld, v.pos).xyz;
+        o.uv = TRANSFORM_TEX(v.uv, _MainTex);
+        return o;
+      }
+
+      // Calculates the grid color at a specified uv position.
+      fixed4 GetGridColor(
+          fixed4 gridColor,
+          float gridSize,
+          float gridLineWidth,
+          float2 uv) {
+        const float Pi = 3.14159;
+        const float GridLineDensity = 3000.0;
+        // Scales a periodic curve (y = cos(x)) to the expected period and amplitude, then clips an
+        // arc out of each period of the curve.
+        float density = GridLineDensity / gridSize;
+        float2 curve = abs(cos(gridSize * Pi * (uv - .5)));
+        float2 clip = saturate((curve - 1) * density / gridLineWidth + 1);
```

Comment:
This formula is used to clip the cos curve to form the line. I don't know if the clipped line width is exact the value of gridLineWidth. Or at least the line width 2 shall be double of the line width 1. Could this formula guarantee this relationship?

Reply:
No, the width of the clipped arc is not strictly proportional to the LineWidth argument. But the result is good enough.

For an exact width of lin rendering, we can use the modulo operation of integers. I dislike the modulo solution because: 1) integer divide and modulo is rather slow when running in GPU. 2) a simple modulo solution results in very sharp and ugly edges in line-rendering. While a cos(x) solution automatically results in a line rendering with gradient edges - since the strength signal is represented by the clipped arc.

Overall I think it's not necessary to take more cost to get a strict proportional line width here.